src/log_batch.rs

```diff
@@ -485,115 +472,128 @@ where
         self.items.is_empty()
     }
 
-    pub fn from_bytes(
-        buf: &mut SliceReader<'_>,
-        file_id: FileId,
-        // The offset of the batch from its log file.
-        base_offset: u64,
-    ) -> Result<Option<LogBatch<M>>> {
-        if buf.is_empty() {
-            return Ok(None);
-        }
-        if buf.len() < BATCH_MIN_SIZE {
-            return Err(Error::TooShort);
-        }
-
-        let header = codec::decode_u64(buf)? as usize;
-        let batch_len = header >> 8;
-        if batch_len > buf.len() {
-            return Err(Error::TooShort);
-        }
-        let batch_type = CompressionType::from_byte(header as u8);
-        test_batch_checksum(&buf[..batch_len])?;
-        trace!("decoding log batch({}, {:?})", batch_len, batch_type);
-
-        let decompressed = match batch_type {
-            CompressionType::None => Cow::Borrowed(&buf[..(batch_len - CHECKSUM_LEN)]),
-            CompressionType::Lz4 => Cow::Owned(decompress(&buf[..batch_len - CHECKSUM_LEN])),
-        };
+    pub fn parse_header(buf: &mut SliceReader<'_>) -> Result<(usize, u64, CompressionType)> {
+        assert_eq!(buf.len(), HEADER_LEN);
+        let mut len = codec::decode_u64(buf)? as usize;
+        let offset = codec::decode_u64(buf)?;
+        let compression_type = CompressionType::from_byte(len as u8);
+        len >>= 8;
+        Ok((len, offset, compression_type))
+    }
 
-        let mut reader: SliceReader = decompressed.borrow();
-        let content_len = reader.len() + HEADER_LEN; // For its header.
-
-        let mut items_count = codec::decode_var_u64(&mut reader)? as usize;
-        assert!(items_count > 0 && !reader.is_empty());
-        let mut log_batch = LogBatch::with_capacity(items_count);
-        let mut entries_size = 0;
-        while items_count > 0 {
-            let content_offset = (content_len - reader.len()) as u64;
-            let item = LogItem::from_bytes(
-                &mut reader,
-                file_id,
-                base_offset,
-                content_offset,
-                &mut entries_size,
-            )?;
-            log_batch.items_size += item.compute_size();
+    pub fn from_footer(
```

Comment:
I feel like we shouldn't allow partially filled `LogBatch`.
Could we do this instead: `LogBatch = (Vec<Entry>, LogItemBatch), LogItemBatch = (Vec<LogItem>), LogItemContent::Entries = (Vec<EntryIndex>)`, so here `from_footer` can belong to `LogItemBatch`.